endpoints/put_test.go

```diff
@@ -33,189 +35,240 @@ import (
+}
+
+type expectedOut struct {
+	PutOutput *PutResponse `json:"put_response"`
+	GetOutput string       `json:"get_response"`
+	Code      int          `json:"code"`
+	ErrorMsg  string       `json:"expected_error_message"`
 }
 
 type logEntry struct {
 	Message string `json:"message"`
 	Level   uint32 `json:"level"`
 }
 
-type testConfig struct {
-	AllowSettingKeys bool `json:"allow_setting_keys"`
-	MaxSizeBytes     int  `json:"max_size_bytes"`
-	MaxNumValues     int  `json:"max_num_values"`
-	MaxTTLSeconds    int  `json:"max_ttl_seconds"`
+type hostConfig struct {
+	AllowSettingKeys bool        `json:"allow_setting_keys"`
+	MaxSizeBytes     int         `json:"max_size_bytes"`
+	MaxNumValues     int         `json:"max_num_values"`
+	MaxTTLSeconds    int         `json:"max_ttl_seconds"`
+	FakeBackend      fakeBackend `json:"fake_backend"`
+}
+
+type fakeBackend struct {
+	ErrorMsg   string       `json:"throw_error_message"`
+	ReturnBool bool         `json:"throw_bool"`
+	StoredData []storedData `json:"stored_data"`
+	Type       string       `json:"storage_type"`
+}
+
+type storedData struct {
+	Key   string          `json:"key"`
+	Value json.RawMessage `json:"value"`
+}
+
+// newTestBackend returns an error-prone backend when a non-empty mockCfg.ErrorMsg string
+// is provided. With an empty mockCfg.ErrorMsg, it returns a well-behaved mock backend.
+func newTestBackend(fb fakeBackend, ttl int) backends.Backend {
+	var mb backends.Backend
+	var backendType config.BackendType = config.BackendType(fb.Type)
+
+	if len(fb.ErrorMsg) > 0 {
+		// "errorProne" backend
+		switch backendType {
+		case config.BackendCassandra:
+			mb = backends.NewMockCassandraBackend(
+				ttl,
+				&backends.ErrorProneCassandraClient{
+					Applied:     fb.ReturnBool,
+					ServerError: errors.New(fb.ErrorMsg),
+				},
+			)
+		case config.BackendMemcache:
+			mb = backends.NewMockMemcacheBackend(&backends.ErrorProneMemcache{ServerError: errors.New(fb.ErrorMsg)})
+		case config.BackendAerospike:
+			mb = backends.NewMockAerospikeBackend(
+				&backends.ErrorProneAerospikeClient{
+					ServerError: fb.ErrorMsg,
+				},
+			)
+		case config.BackendRedis:
+			var serverErr error
+			if fb.ErrorMsg == "redis: nil" {
+				serverErr = redis.Nil
+			} else {
+				serverErr = errors.New(fb.ErrorMsg)
+			}
+			mb = backends.NewFakeRedisBackend(
+				backends.FakeRedisClient{
+					StoredData:  copyStoredData(fb.StoredData),
+					ServerError: serverErr,
+					Success:     fb.ReturnBool,
+				},
+			)
+		default:
+			mb = backends.NewErrorResponseMemoryBackend()
+		}
+		return mb
+	}
+
+	// Well-behaved mock backend
+	switch backendType {
+	case config.BackendCassandra:
+		mb = backends.NewMockCassandraBackend(
+			ttl,
+			&backends.GoodCassandraClient{
+				StoredData: copyStoredData(fb.StoredData),
+			},
+		)
+	case config.BackendMemcache:
+		mb = backends.NewMockMemcacheBackend(
+			&backends.GoodMemcache{
+				copyStoredData(fb.StoredData),
```

Comment:
Nitpick: can you specify the key name `StoredData` here?

Reply:
Done